compiler/rustc_codegen_llvm/src/asm.rs

```diff
@@ -874,3 +903,35 @@ fn llvm_fixup_output_type(
         _ => layout.llvm_type(cx),
     }
 }
+
+// Valid llvm symbols: https://llvm.org/docs/AMDGPUOperandSyntax.html#symbols
+// First char is [a-zA-Z_.]
+fn llvm_ident_start(c: char) -> bool {
+    ('a'..='z').contains(&c) || ('A'..='Z').contains(&c) || '_' == c || '.' == c
+}
+
+// All subsequent chars are [a-zA-Z0-9_$.@]
```

Comment:
I would prefer if we disallowed `@` and `$` in symbol names. We don't need to support *all* symbol names here, the user is free to shoot themselves in the foot if they choose to use a symbol name with weird characters.

Reply:
> I would prefer if we disallowed @ and $ in symbol names.

That should happen elsewhere, possibly in the same location parsing for formatting specifiers happens. Should also be a concern for a separate PR.

Reply:
My main concern here is that these may not be valid characters in symbol names for all targets and may have other special meanings on those targets.

Reply:
Sure, I agree. I still think this is the the wrong place in the codebase to enforce what characters we allow in symbols. Probably should happen as a separate PR too.

Reply:
We don't actually prevent the user from using such labels. We just don't provide automatic rewriting into local labels for them.

Also we're not handling all possible label definitions. For example labels can be defined after a `;` on targets where that is a statement separator.

We're not trying to handle all possible cases here since that would require us to effectively parse the asm, which we really want to leave to LLVM. We're just trying to make the most common types of labels work correctly.

Reply:
> We don't actually prevent the user from using such labels. We just don't provide automatic rewriting into local labels for them.

Aha, I see the distinction you're trying to make now.

> We're just trying to make the most common types of labels work correctly.

That seems like a recipe for a pretty poor experience with a very non-obvious cliff at which things may start failing in non-obvious ways. For instance, user with passing familiarity with Rust's `asm!` labels being local could write something like this:

```
#[inline] // not always
pub fn peach() {
    unsafe {
        asm!("per$$imon: jmp per$$imon")
    }
}
```

and it might very well work out in basic tests, but start failing once the function is released to the field. If we do go the route of making things automagic, its better they're magic all the time, not only sometimes.

> that would require us to effectively parse the asm, which we really want to leave to LLVM

I believe just lexing would be sufficient, but I do agree that we'd rather not do that at all.

Reply:
At the very least I would like `@` to be removed from the set of allowed characters for symbols. `@` is used as a symbol modifier in x86 intel syntax: `sym@gotpcrel`.

Reply:
I want to confirm what we want to accept as a valid identifier for the purposes of *replacing labels*.  Denying symbols outright should be properly implemented somewhere else, preferably with a compiler warning or error.  This code can reflect what symbols are accepted, but it will ignore other things.  For example, `@foobar:` will be replaced with `@${:private}foobar${:uid}` if we choose to not acknowledge `@` in labels.

Is the set of symbols we will accept in a label `[a-zA-Z0-9_$.]`?

Additionally, do we want to accept labels that are strings like nagisa commented [here](https://github.com/rust-lang/rust/pull/81570#discussion_r568536405)?

Reply:
I think that for label renaming we should only support `[a-zA-Z0-9_.]`, and only plain labels not quoted ones.